frontend/src/metabase-lib/v1/expressions/pratt/compiler.unit.spec.ts

```diff
@@ -83,59 +90,115 @@ describe("pratt/compiler", () => {
 
     // note, changed from original to accommodate validation of substring args
     it(`Seed 10099: CONtAinS ( [OF4wuV], SUbstriNG("_", 1, lENGtH("s Mfg7" ) ) )`, () => {
-      const { oracle, compiled } = compare(
+      const compiled = expr(
         `CONtAinS ( [OF4wuV], SUbstriNG("_", 1, lENGtH("s Mfg7" ) ) )`,
-        "boolean",
-        { throwOnError: true, resolverPass: false },
       );
-      expect(compiled).toEqual(oracle);
+      expect(compiled).toEqual([
+        "contains",
+        ["dimension", "OF4wuV"],
+        ["substring", "_", 1, ["length", "s Mfg7"]],
+      ]);
     });
 
     it(`Seed 10092:  NOT ( NOT   (isNUll ([T0q → n_M_O])))`, () => {
-      const { oracle, compiled } = compare(
-        ` NOT ( NOT   (isNUll ([T0q → n_M_O])))`,
-        "expression",
-        {
-          throwOnError: true,
-          resolverPass: false,
-        },
-      );
-      expect(compiled).toEqual(oracle);
+      const compiled = expr(` NOT ( NOT   (isNUll ([T0q → n_M_O])))`);
+      expect(compiled).toEqual([
+        "not",
+        ["not", ["is-null", ["dimension", "T0q → n_M_O"]]],
+      ]);
     });
 
     it(`Seed 10082: SUbstriNg( cOncat("BaK2    ", [__m_4], rTrim(coNcAt (replACE ([Av5Wtbz], regeXextRACt( [_1I → g], "H NVB84_"), rEGexextract ( [__8], " _ 2" )  ) , SUbStRiNG("qb0  ", (power( LeNgTh ( rtrim ( "YyCe_2" )) * 0e+77, 1 ) ), 1 + 0e-54 / 374719e-64) , cOncaT( "    F9 _O", "_a5_", " 5 _U_ ", " bE", rEPlACe (BXj3O, " ", [Z → X9]) ) )  )  ), (1), log ( 1E-26 )  )`, () => {
-      const { oracle, compiled } = compare(
+      const compiled = expr(
         `SUbstriNg( cOncat("BaK2    ", [__m_4], rTrim(coNcAt (replACE ([Av5Wtbz], regeXextRACt( [_1I → g], "H NVB84_"), rEGexextract ( [__8], " _ 2" )  ) , SUbStRiNG("qb0  ", (power( LeNgTh ( rtrim ( "YyCe_2" )) * 0e+77, 1 ) ), 1 + 0e-54 / 374719e-64) , cOncaT( "    F9 _O", "_a5_", " 5 _U_ ", " bE", rEPlACe (BXj3O, " ", [Z → X9]) ) )  )  ), (1), log ( 1E-26 )  )`,
-        "expression",
-        { throwOnError: true, resolverPass: false },
       );
-      expect(compiled).toEqual(oracle);
+      expect(compiled).toEqual([
```

Comment:
To remove the need for `oracle`, which was the only caller of `recursive-parser` left, the test cases are hardcoded here.